data-products/src/main/scala/org/sunbird/analytics/util/TextBookUtils.scala

```diff
@@ -0,0 +1,183 @@
+                             subject: String, status: String, createdOn: String, lastUpdatedOn: String, totalContentLinked: Integer,
+                             totalQRLinked: Integer, totalQRNotLinked: Integer, leafNodesCount: Integer, leafNodeUnlinked: Integer)
+
+case class dceTextbookData(channel: String, identifier: String, name: String, medium: String, gradeLevel:String, subject: String,
+                             createdOn: String, lastUpdatedOn: String, totalQRCodes: Integer, contentLinkedQR: Integer,
+                             withoutContentQR: Integer, withoutContentT1: Integer, withoutContentT2: Integer)
+
+object TextBookUtils {
+
+  def getTextBooks(config: Map[String, AnyRef], restUtil: HTTPClient): List[TextBookInfo] = {
+    val apiURL = Constants.COMPOSITE_SEARCH_URL
+    val request = JSONUtils.serialize(config.get("esConfig").get)
+    val response = restUtil.post[TextBookDetails](apiURL, request).result.content
+    response
+  }
+
+  def getTextbookHierarchy(textbookInfo: List[TextBookInfo],tenantInfo: RDD[TenantInfo],restUtil: HTTPClient)(implicit sc: SparkContext): (RDD[FinalOutput]) = {
+    var etbTextBookReport = List[etbTextbookData]()
+    var dceTextBookReport = List[dceTextbookData]()
+
+    textbookInfo.map(textbook => {
+      var apiUrl = AppConf.getConfig("hierarchy.search.api.url")+AppConf.getConfig("hierarchy.search.api.path")
+      if(textbook.status == "Live") { apiUrl = apiUrl+textbook.identifier }
+      else { apiUrl = apiUrl+textbook.identifier+"?mode=edit" }
+
+      val response = restUtil.get[ContentDetails](apiUrl)
+
+      if(response.params.status=="successful") {
+        val data = response.result.content
+
+        val etbTextBookResponse = generateETBTextbookReport(data)
+        etbTextBookReport = etbTextBookReport ++ etbTextBookResponse
+
+        val dceTextBookResponse = generateDCETextbookReport(data)
+        dceTextBookReport = dceTextBookReport ++ dceTextBookResponse
+      }
+    })
+    val tenantRDD = tenantInfo.map(e=>(e.id,e))
+
+    val etbTextBook = sc.parallelize(etbTextBookReport).map(e=>(e.channel,e))
+    val etb=etbTextbookData("","","","","","","","","",0,0,0,0,0)
+    val etbTextBookRDD = etbTextBook.fullOuterJoin(tenantRDD).map(etbReport => {
+
+      etbTextbookReport(etbReport._2._2.getOrElse(TenantInfo("","unknown")).slug, etbReport._2._1.getOrElse(etb).identifier,
+        etbReport._2._1.getOrElse(etb).name,etbReport._2._1.getOrElse(etb).medium,etbReport._2._1.getOrElse(etb).gradeLevel,
+        etbReport._2._1.getOrElse(etb).subject,etbReport._2._1.getOrElse(etb).status,etbReport._2._1.getOrElse(etb).createdOn,etbReport._2._1.getOrElse(etb).lastUpdatedOn,
+        etbReport._2._1.getOrElse(etb).totalContentLinked,etbReport._2._1.getOrElse(etb).totalQRLinked,etbReport._2._1.getOrElse(etb).totalQRNotLinked,
+        etbReport._2._1.getOrElse(etb).leafNodesCount,etbReport._2._1.getOrElse(etb).leafNodeUnlinked,"ETB_textbook_data")
```

Comment:
why do we need an "etb" prefix for every variable? just use the generic variables textbook and report as the class name already contains ETB in it

Reply:
Done